File: app/src/main/java/com/mdkashif/alarm/base/BaseActivity.java
package com.mdkashif.alarm.base;


import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.content.Context;
import android.location.Location;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Toast;

import com.mdkashif.alarm.R;
import com.mdkashif.alarm.battery.BatteryReceiver;
import com.mdkashif.alarm.custom.CustomProgressDialog;
import com.mdkashif.alarm.prayer.geocoder.GetCurrentLocation;
import com.mdkashif.alarm.prayer.geocoder.GetLocationAddress;
import com.mdkashif.alarm.utils.AppDatabase;

public class BaseActivity extends AppCompatActivity {
    CustomProgressDialog progressDialog;
    View parentLayout;
    static public String city,country;
    BatteryReceiver batteryReceiver=BatteryReceiver.getInstance();
    public String tag = "universal_alarm_";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        parentLayout = findViewById(android.R.id.content);
        Location location = GetCurrentLocation.getInstance().findLocation(getApplicationContext());
        progressDialog = new CustomProgressDialog(BaseActivity.this);
        if(location!=null) {
            double latitude = location.getLatitude();
            double longitude = location.getLongitude();
            GetLocationAddress.getAddressFromLocation(latitude, longitude,
                    getApplicationContext(), new GeocoderHandler());
        }

    }

    public void showLoader() {
        progressDialog.show();
    }

    public void hideLoader() {
        progressDialog.dismiss();
    }

    public void showToast(String message) {
        Toast.makeText(BaseActivity.this, message + "", Toast.LENGTH_SHORT).show();
    }

    public void showSnackbar(String message,Context context) {
        Snackbar.make(parentLayout.findViewById(android.R.id.content),
                message, Snackbar.LENGTH_LONG).show();
    }

    public boolean isOnline() {
        ConnectivityManager cm = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo netInfo = cm.getActiveNetworkInfo();
        if (netInfo != null && netInfo.isConnectedOrConnecting() && cm.getActiveNetworkInfo().isAvailable() && cm.getActiveNetworkInfo().isConnected()) {
            return true;
        }
        return false;
    }

    public boolean isBlank(String strValue) {
        if (strValue == null || strValue.equals(""))
            return true;
        else
            return false;
    }

    public void replaceFragment(Fragment fragment, String tag, boolean isAddTobackStack) {
        FragmentManager fm = getSupportFragmentManager();
        FragmentTransaction ft = fm.beginTransaction();
        ft.setCustomAnimations( R.animator.slide_up, 0, 0, R.animator.slide_down);
        ft.replace(R.id.fl_main_content, fragment, tag);
        if (isAddTobackStack) {
            ft.addToBackStack(tag);
        }
        ft.commit();
    }

    private static class GeocoderHandler extends Handler {
        @Override
        public void handleMessage(Message message) {
            switch (message.what) {
                case 1:
                    Bundle bundle = message.getData();
                    city = bundle.getString("city");
                    country = bundle.getString("country");
                    break;
                default:
                    city = "";
                    country = "";
            }
        }
    }

    @Override
    protected void onDestroy() {
        AppDatabase.destroyInstance();
        super.onDestroy();
    }

}

File: app/src/main/java/com/mdkashif/alarm/activities/BaseActivity.kt
package com.mdkashif.alarm.activities

import android.content.Context
import android.content.res.Configuration
import android.net.ConnectivityManager
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.view.View
import android.view.animation.*
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import com.mdkashif.alarm.R
import com.mdkashif.alarm.alarm.miscellaneous.AlarmListAdapter
import com.mdkashif.alarm.alarm.prayer.geocoder.GetCurrentLocation
import com.mdkashif.alarm.alarm.prayer.geocoder.GetLocationAddress
import com.mdkashif.alarm.custom.CustomProgressDialog
import com.mdkashif.alarm.custom.SwipeToDeleteCallback
import com.mdkashif.alarm.utils.AppConstants
import com.mdkashif.alarm.utils.db.AppDatabase


open class BaseActivity : AppCompatActivity() {
    private var progressDialog: CustomProgressDialog? = null
    private var parentLayout: View? = null
//    private val batteryReceiver = BatteryReceiver.getInstance()
    lateinit var appDatabase: AppDatabase

    val isOnline: Boolean
        get() {
            val cm = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            val netInfo = cm.activeNetworkInfo
            return netInfo != null && netInfo.isConnectedOrConnecting && cm.activeNetworkInfo.isAvailable && cm.activeNetworkInfo.isConnected
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        parentLayout = findViewById(android.R.id.content)
        val location = GetCurrentLocation.getInstance().findLocation(applicationContext)
        progressDialog = CustomProgressDialog(this@BaseActivity)
        if (location != null) {
            val latitude = location.latitude
            val longitude = location.longitude
            GetLocationAddress.getAddressFromLocation(latitude, longitude,
                    applicationContext, GeocoderHandler())
        }

        appDatabase = AppDatabase.getAppDatabase(applicationContext)
    }

    fun showLoader() {
        progressDialog!!.show()
    }

    fun hideLoader() {
        progressDialog!!.dismiss()
    }

    fun showToast(message: String) {
        Toast.makeText(this@BaseActivity, message + "", Toast.LENGTH_SHORT).show()
    }

    fun showSnackBar(message: String) {
        Snackbar.make(parentLayout!!.findViewById(android.R.id.content),
                message, Snackbar.LENGTH_LONG).show()
    }

    fun isBlank(strValue: String?): Boolean {
        return strValue == null || strValue == ""
    }

    fun replaceFragment(fragment: Fragment, tag: String, isAddToBackStack: Boolean) {
        val fm = supportFragmentManager
        val ft = fm.beginTransaction()
        ft.setCustomAnimations(R.animator.slide_in,
                R.animator.slide_out)
        ft.replace(R.id.flContainer, fragment, tag)
        if (isAddToBackStack) {
            ft.addToBackStack(tag)
        }
        ft.commit()
    }

    fun setRVSlideInLeftAnimation(view: RecyclerView){
        val set = AnimationSet(true)
        var animation: Animation = AlphaAnimation(0.0f, 1.0f)
        animation.duration = 100
        set.addAnimation(animation)
        animation = TranslateAnimation(
                Animation.RELATIVE_TO_SELF, 1.0f, Animation.RELATIVE_TO_SELF, 0.0f,
                Animation.RELATIVE_TO_SELF, 0.0f, Animation.RELATIVE_TO_SELF, 0.0f
        )
        animation.duration=800
        set.addAnimation(animation)

        val controller = LayoutAnimationController(set, 0.5f)
        view.layoutAnimation = controller
    }

    fun enableSwipeToDeleteAndUndo(mAdapter : AlarmListAdapter, mRecyclerView: RecyclerView) {
        val swipeToDeleteCallback = object : SwipeToDeleteCallback(this) {
            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, i: Int) {
                val position = viewHolder.adapterPosition
                val item = mAdapter.data[position]

                mAdapter.removeItem(position)

                val snackBar = Snackbar
                        .make(parentLayout!!.findViewById(android.R.id.content),
                                "Alarm removed", Snackbar.LENGTH_LONG)
                snackBar.setAction("UNDO") {
                    mAdapter.restoreItem(item, position)
                    mRecyclerView.scrollToPosition(position)
                }
                snackBar.setActionTextColor(resources.getColor(R.color.gray))
                snackBar.show()
            }
        }

        val itemTouchHelper = ItemTouchHelper(swipeToDeleteCallback)
        itemTouchHelper.attachToRecyclerView(mRecyclerView)
    }

    private class GeocoderHandler : Handler() {
        override fun handleMessage(message: Message) {
            when (message.what) {
                1 -> {
                    val bundle = message.data
                    city = bundle.getString("city")
                    country = bundle.getString("country")
                }
                else -> {
                    city = ""
                    country = ""
                }
            }
        }
    }

    fun returnDbInstance():AppDatabase{
        return appDatabase
    }

    fun detectThemeAuto(): String{
        var autoTheme = ""
        when (resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK) {
            Configuration.UI_MODE_NIGHT_NO-> autoTheme= AppConstants.themeLight

            Configuration.UI_MODE_NIGHT_YES-> autoTheme= AppConstants.themeDark

            Configuration.UI_MODE_NIGHT_UNDEFINED-> autoTheme= "undefined"
        }
        return autoTheme
    }

    override fun onDestroy() {
        AppDatabase.destroyInstance()
        super.onDestroy()
    }

    companion object {
        var city: String? = null
        var country: String? = null
    }

}
